File: SPOJ/PT07Y.py
def dfs(node):
    if visited[node]:
        return 
    visited[node] = 1 
    for nbr in graph[node]:
        dfs(nbr)

n,m = map(int,input().split())
graph = dict()
for i in range(1,n+1):
    graph[i] = []
visited = [0 for i in range(n+1)]
for _ in range(m):
    u,v = map(int,input().split())
    graph[u].append(v)
    graph[v].append(u)
if m != n-1:
    print('NO')
else: 
    dfs(1)
    isTree = True
    for k in range(1,n+1):
        if visited[k]==0:
            isTree = False
            break
    if isTree:
        print('YES')
    else:
        print('NO')

File: SPOJ/PT07Y.cpp
#include<bits/stdc++.h>
using namespace std;

void dfs(int node, unordered_map<int, vector<int>> &graph, vector<int> &visited){
    if(visited[node]==1) return;
    visited[node] = 1;
    for(int nbr :  graph[node]){
        dfs(nbr,graph,visited);
    }
}

int main(){
    int n,m;
    cin >> n >> m;
    unordered_map<int,vector<int>> graph;
    vector<int> visited(n+1,0);
    for(int i = 1; i <= n; i++){
        graph[i] = {};
    }
    for(int i = 1; i <= m; i++){
        int u,v;
        cin >> u >> v;
        graph[u].emplace_back(v);
        graph[v].emplace_back(u);
    }
    if(m != n-1){
        cout << "NO" << endl;
    }else{
        dfs(1,graph,visited);
        int isTree = 1;
        for(int i = 1; i <= n; i++){
            if(visited[i]==0){
                isTree = 0;
                break;
            }
        }
        if(isTree==1){
            cout << "YES" << endl;
        }else{
            cout << "NO" <<  endl;
        }
    }
}